waiter/src/waiter/scheduler.clj

```diff
@@ -665,6 +666,46 @@
                         service
                         (assoc instances :active-instances active-instances))))))))
 
+(defn start-new-services-daemon
+  "Listens on token-metric-chan-mult and determines whether to trigger start new services for a token. Only the leader
+  router starts new services. Tokens with new last-request-time pointing to a service descriptor that does not exist will
+  have the latest service started.
+
+  Returns a map with keys:
+  :process-token-event-ch is a channel where messages are pushed to it when a service was started for a token event."
+  [retrieve-latest-descriptor-fn service-exists? kv-store token-metric-chan-mult start-new-service-fn leader?-fn]
+  (cid/with-correlation-id
+    "start-new-services-goroutine"
+    (let [correlation-id (cid/get-correlation-id)
+          process-token-event-ch-buffer (async/sliding-buffer 1000)
+          process-events-fn
+          (fn process-events [events]
+            (filter
+              (fn latest-service-does-not-exist?
+                [{:keys [token]}]
+                (when (leader?-fn)
+                  (let [{:strs [run-as-user]}
+                        (sd/token->service-parameter-template kv-store token :error-on-missing false)
+                        {:keys [service-id] :as latest-descriptor}
+                        (retrieve-latest-descriptor-fn run-as-user token)
+                        service-does-not-exist? (not (service-exists? service-id))]
+                    (when service-does-not-exist?
+                      (cid/cinfo correlation-id "starting" {:service-id (get latest-descriptor :service-id)})
+                      (start-new-service-fn latest-descriptor))
+                    service-does-not-exist?)))
+              events))
+          process-token-event-ch
+          (async/chan
+            process-token-event-ch-buffer
+            (comp
+              (map process-events-fn)
```

Comment:
Typically the transform does not have side-effects. We should refactor this later into a go-block to read data from the channel and invoke `process-events`.

Reply:
sgtm